src/Swarm/Game/State.hs

```diff
@@ -1233,23 +1241,24 @@ scenarioToGameState scenario (LaunchParams (Identity userSeed) (Identity toRun))
 
 -- | Take a world description, parsed from a scenario file, and turn
 --   it into a list of located robots and a world function.
-buildWorld :: EntityMap -> WorldDescription -> ([IndexedTRobot], Seed -> WorldFun Int Entity)
-buildWorld em WorldDescription {..} = (robots worldName, first fromEnum . wf)
+buildWorld :: WorldDescription -> ([IndexedTRobot], Seed -> WorldFun Int Entity)
+buildWorld WorldDescription {..} = (robots worldName, first fromEnum . wf)
  where
   rs = fromIntegral $ length area
   cs = fromIntegral $ length (head area)
   Coords (ulr, ulc) = locToCoords ul
 
-  worldGrid :: [[(TerrainType, Maybe Entity)]]
+  worldGrid :: [[(TerrainType, Erasable Entity)]]
   worldGrid = (map . map) (cellTerrain &&& cellEntity) area
 
-  worldArray :: Array (Int32, Int32) (TerrainType, Maybe Entity)
+  worldArray :: Array (Int32, Int32) (TerrainType, Erasable Entity)
   worldArray = listArray ((ulr, ulc), (ulr + rs - 1, ulc + cs - 1)) (concat worldGrid)
 
-  wf = case defaultTerrain of
-    Nothing ->
-      (if offsetOrigin then findGoodOrigin else id) . testWorld2FromArray em worldArray
-    Just (Cell t e _) -> const (worldFunFromArray worldArray (t, e))
+  dslWF, arrayWF :: Seed -> WorldFun TerrainType Entity
+  dslWF = maybe mempty (((if offsetOrigin then findGoodOrigin else id) .) . runTTerm) worldProg
```

Comment:
[`applyWhen`](https://hackage.haskell.org/package/base-4.18.0.0/docs/Data-Function.html#v:applyWhen)?

Reply:
Cool, I didn't know about `applyWhen`!  Though it looks like it has only been in `base` since version 4.18, so if we want Swarm to build with older versions of GHC/`base` we can't use it yet.

...oh, wait!  We can depend on the version from `base-compat`, that's exactly what `base-compat` is for.

Reply:
Well, actually, the LTS version we are currently pegged to has `base-compat-0.12.2`, which does not contain `applyWhen`.  I guess we'll have to wait a little while yet.

Reply:
For now I stuck a version of `applyWhen` in `Swarm.Util` with a note to switch to the version from `base-compat` once it becomes available.